tests/bsim/bluetooth/samples/battery_service/src/central_test.c

```diff
@@ -0,0 +1,441 @@
+}
+
+static void test_bas_central_tick(bs_time_t HW_device_time)
+{
+	/*
+	 * If in WAIT_TIME seconds the testcase did not already pass
+	 * (and finish) we consider it failed
+	 */
+	if (bst_result != Passed) {
+		FAIL("test_bas_central failed (not passed after %i seconds)\n", WAIT_TIME);
+	}
+}
+
+/* Callback for handling Battery Level Notifications */
+static uint8_t battery_level_notify_cb(struct bt_conn *conn,
+				       struct bt_gatt_subscribe_params *params, const void *data,
+				       uint16_t length)
+{
+	if (data) {
+		printk("[NOTIFICATION] BAS Battery Level: %d%%\n", *(const uint8_t *)data);
+	} else {
+		printk("Battery Level Notifications disabled\n");
+	}
+	return BT_GATT_ITER_CONTINUE;
+}
+
+/* Callback for handling Battery Level Read Response */
+static uint8_t battery_level_read_cb(struct bt_conn *conn, uint8_t err,
+				     struct bt_gatt_read_params *params, const void *data,
+				     uint16_t length)
+{
+	if (err) {
+		printk("Failed to read Battery Level (err %u)\n", err);
+		return BT_GATT_ITER_STOP;
+	}
+
+	if (data) {
+		printk("[READ] BAS Battery Level: %d%%\n", *(const uint8_t *)data);
+	}
+
+	return BT_GATT_ITER_STOP;
+}
+
+static unsigned char battery_level_status_indicate_cb(struct bt_conn *conn,
+						      struct bt_gatt_subscribe_params *params,
+						      const void *data, uint16_t length)
+{
+	static int ind_received;
+
+	if (!data) {
+		printk("bas level status indication disabled\n");
+	} else {
+		printk("[INDICATION] BAS Battery Level Status: ");
+		for (int i = 0; i < length; i++) {
+			printk("%02x ", ((uint8_t *)data)[i]);
+		}
+		printk("\n");
+
+		if (ind_received++ > BAS_BLS_IND_RECEIVED_COUNT) {
+			PASS();
+		}
+	}
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static uint8_t battery_level_status_notify_cb(struct bt_conn *conn,
+					      struct bt_gatt_subscribe_params *params,
+					      const void *data, uint16_t length)
+{
+	static int notify_count;
+
+	if (!data) {
+		printk("[UNSUBSCRIBED]\n");
+		params->value_handle = 0U;
+		return BT_GATT_ITER_STOP;
+	}
+
+	printk("[NOTIFICATION] BAS Battery Level Status: ");
+	for (int i = 0; i < length; i++) {
+		printk("%02x ", ((uint8_t *)data)[i]);
+	}
+	printk("\n");
+
+	if (notify_count++ >= BAS_BLS_NTF_RECEIVED_COUNT) { /* We consider it passed */
+		PASS();
+	}
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static void read_battery_level(const struct bt_gatt_attr *attr)
+{
+
+	/* Read the battery level after subscribing */
+	static struct bt_gatt_read_params read_params;
+
+	read_params.func = battery_level_read_cb;
+	read_params.handle_count = 1;
+	read_params.single.handle = bt_gatt_attr_get_handle(attr);
+	read_params.single.offset = 0;
+	bt_gatt_read(default_conn, &read_params);
+}
+
+static void subscribe_battery_level(const struct bt_gatt_attr *attr)
+{
+	int err;
+
+	battery_level_notify_params = (struct bt_gatt_subscribe_params){
+		.ccc_handle = bt_gatt_attr_get_handle(attr) + 2,
+		.value_handle = bt_gatt_attr_get_handle(attr) + 1,
```

Comment:
This is an OK assumption in BSIM as Zephyr always does this, but not a general assumption that can be used. 

Suggest to add a comment stating this